### ISSUE TITLE
Some questions about lime

### ISSUE THREAD

**Issue by sym823808458** [deepchem/deepchem#1551]
Hello everyone,
I am new here in deepchem.I found Lime is a great way to explain the molecular fingerprint.But I don't quite understand it. I do know the deeplearning NN separated the molecules into small segments like some functional groups and the Lime is a tool to show the fingerprint.
So my first question is that what is the relationship between the molecule and the fingerprint?
After running the code below,I tried to see 'my_fp'.And I saw some numbers explained by some chemical functional group,like,314: {'C=O'}, 594: {'CC(C)(O)CC[CH]'}.Somehow,I see one thing like'807: {''},',which is an empty column.What does that mean?
More important thing is that I really need to see which fingerprint contributes most in the whole trainingset.In other words,how to see the feature importance?For example,in tox21 database,after the DNN training,maybe some functional group can totally lead to the 'toxic' result,maybe some can lead to 80%'toxic'.How can I see the information like this?
Maybe what I wanna do make no sense in deepchem.More informations of Lime is really needed here.
Thanks for any kind of help.

'''
#!/usr/bin/env python
# coding: utf-8

# In[1]:

from rdkit import Chem
from rdkit.Chem.Draw import IPythonConsole
from IPython.display import SVG
from rdkit.Chem import rdDepictor
from rdkit.Chem.Draw import rdMolDraw2D

import numpy as np
import deepchem as dc
from deepchem.molnet import load_tox21

# Only for debug!
np.random.seed(123)

# Load Tox21 dataset
n_features = 1024
tox21_tasks, tox21_datasets, transformers = load_tox21()
train_dataset, valid_dataset, test_dataset = tox21_datasets

# In[2]:

# Fit models
metric = dc.metrics.Metric(
    dc.metrics.roc_auc_score, np.mean, mode="classification")

nb_epoch = 10
model = dc.models.tensorgraph.fcnet.MultitaskClassifier(
    len(tox21_tasks),
    train_dataset.get_data_shape()[0])

# Fit trained model
model.fit(train_dataset, nb_epoch=nb_epoch)
model.save()

print("Evaluating model")
train_scores = model.evaluate(train_dataset, [metric], transformers)
valid_scores = model.evaluate(valid_dataset, [metric], transformers)

# In[3]:

print("Train scores")
print(train_scores)

print("Validation scores")
print(valid_scores)

# In[4]:

#So LIME as is can work on any problem with a fixed size input vector. 
#It works by computing probability distributions for the individual features and covariance between the features.
#We are going to create an explainer for our data

from lime import lime_tabular
feature_names = ["fp_%s"  % x for x in range(1024)]
explainer = lime_tabular.LimeTabularExplainer(train_dataset.X,
                                              feature_names=feature_names,
                                              categorical_features=feature_names,
                                              class_names=['not toxic', 'toxic'],
                                              discretize_continuous=True)

# In[5]:

# We need a function which takes a 2d numpy array (samples, features) and returns predictions (samples,)
def eval_model(my_model, transformers):
    def eval_closure(x):
        ds = dc.data.NumpyDataset(x, None, None, None)
        # The 0th task is NR-AR
        predictions = model.predict(ds)[:, 0]
        return predictions
    return eval_closure
model_fn = eval_model(model, transformers)

# In[6]:

# We want to investigate a toxic compound
active_id = np.where(test_dataset.y[:,0]==1)[0][0]
print(active_id)
Chem.MolFromSmiles(test_dataset.ids[active_id])

# In[7]:

# this returns an Lime Explainer class
# The explainer contains details for why the model behaved the way it did
exp = explainer.explain_instance(test_dataset.X[active_id], model_fn, num_features=5, top_labels=1)
# If we are in an ipython notebook it can show it to us
exp.show_in_notebook(show_table=True, show_all=False)

# In[8]:

#So the model believes that fingerprint 118, 519 and 301 highly contribute to the molecule’s toxicity. 
#We can reverse our the hash function and look at the fragments that activated those fingerprints for this molecule.

from rdkit import Chem

def fp_mol(mol, fp_length=1024):
    """
    returns: dict of <int:list of string>
        dictionary mapping fingerprint index
        to list of smile string that activated that fingerprint
    """
    d = {}
    feat = dc.feat.CircularFingerprint(sparse=True, smiles=True, size=1024)
    retval = feat._featurize(mol)
    for k, v in retval.items():
        index = k % 1024
        if index not in d:
            d[index] = set()
        d[index].add(v['smiles'])
    return d
# What fragments activated what fingerprints in our active molecule?
my_fp=fp_mol(Chem.MolFromSmiles(test_dataset.ids[active_id]))

# In[9]:

# We can calculate which fragments activate all fingerprint
# indexes throughout our entire training set
all_train_fps = {}
X = train_dataset.X
ids = train_dataset.ids
for i in range(len(X)):
    d = fp_mol(Chem.MolFromSmiles(ids[i]))
    for k, v in d.items():
        if k not in all_train_fps:
            all_train_fps[k] = set()
        all_train_fps[k].update(v)

# In[10]:

# We can visualize which fingerprints our model declared toxic for the
# active molecule we are investigating
Chem.MolFromSmiles(list(my_fp[118])[0])

# In[11]:

Chem.MolFromSmiles(list(my_fp[519])[0])

# In[12]:

Chem.MolFromSmiles(list(my_fp[301])[0])

# In[13]:

# We can also see what fragments are missing by investigating the training set
# According to our explanation having one of these fragments would make our molecule more likely to be toxic
Chem.MolFromSmiles(list(all_train_fps[381])[0])

# In[14]:

Chem.MolFromSmiles(list(all_train_fps[381])[1])

# In[15]:

Chem.MolFromSmiles(list(all_train_fps[381])[4])

# In[16]:

Chem.MolFromSmiles(list(all_train_fps[875])[0])

# In[17]:

Chem.MolFromSmiles(list(all_train_fps[875])[2])

# In[18]:

Chem.MolFromSmiles(list(all_train_fps[875])[4])
#Using LIME on fragment based models can give you intuition over which fragments are contributing 
#to your response variable in a linear fashion.
'''

**Comment by rbharath** [comment]
Sorry, this issue slipped through my inbox. If you're still having issues here, would you mind clarifying a bit more? I'm not sure which part of lime you're having issues with
